### PR TITLE
Dev blendmodes class

### PR DESCRIPTION
**Pull request by ivanpopelyshev**
Better version of #4033 

1. Takes care of premultiplied alpha
2. Allows to create your own blendmodes

```
var gl = WebGLRenderingContext;
var blend = new PIXI.BlendMode(gl.ZERO, gl.ONE_MINUS_SRC_ALPHA, gl.ONE, gl.ONE_MINUS_SRC_ALPHA); //RGB=000

sprite.blendMode = blend;
```

3. side effect: "renderer.setBlendMode" and "renderer.state.setBlendMode" aren't accepting numbers anymore. But you still can use "sprite.blendMode = 1" because of property fallback.

### CODE REVIEW

File: src/core/BlendMode.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+        /**
+         * non-premultipled and premultiplied versions.
+         *
+         * @member {Array<PIXI.BlendMode>}
+         */
+        this.npm = [this, this];
+
+        /**
+         * The height of the sprite (this is initially set by the texture)
```
Comment (andrewstart) [inline src/core/BlendMode.js]: I think you forgot to change some docs here.

Comment (ivanpopelyshev) [inline src/core/BlendMode.js]: Lol :) fixed.

**Comment by ivanpopelyshev** [comment]
Note for users who save/load sprites/meshes: 

```
sprite.blendMode.code; // <- this one is a number
PIXI.BlendMode.values[code]; // <-- that's how you get blendmode object by code
```

**Comment by englercj** [comment]
I'm not sure that this level of refactoring is a good idea in v4, and I'm not entirely happy with the API it produces. It is pretty complicated.

**Comment by ivanpopelyshev** [comment]
@englercj well, you asked for classes ;) What if @GoodBoyDigital takes it into v5, and we merge previous PR in v4.

**Comment by mnori** [comment]
Anyone know if this fixes the following?

https://codepen.io/wildtype/pen/NjeRMN

**Comment by englercj** [comment]
> well, you asked for classes ;)

I didn't though, I said it was fine and that I would refactor it into classes-like structure for v5.

**Comment by bigtimebuddy** [comment]
@ivanpopelyshev can this be closed because we merge the other blendmodes pr?

**Comment by ivanpopelyshev** [comment]
@bigtimebuddy this one yes, but I'll make it against v5 ("next" branch)

**Comment by englercj** [comment]
> this one yes, but I'll make it against v5 ("next" branch)

I would hold off on that for a bit, I have a lot of local changes that may invalidate that need.

**Comment by ivanpopelyshev** [comment]
@englercj ok ;)

**Comment by ivanpopelyshev** [comment]
This will be rebased against v5, and I need this PR to be visible for people who want to test it in v4.

**Comment by bigtimebuddy** [comment]
Hey @ivanpopelyshev, we will tackle this in v5 which is rolling right now on the **next** branch, but no major refactors for v4. I'm closing for now, when you are ready to add again to v5, let's re-open or resubmit.

**Pull request closed without merging**